File: 3-Longest Substring Without Repeating Characters.py
class Solution:
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        max_len, start, char_dict = 0, 0, {}
        for i, char in enumerate(s, start=1):  # enumerate这里的start是从1开始循环，而不是0
            if char_dict.get(char, -1) >= start:  # 这里的dict.get(char, -1)中的-1的意思是当没有得到的时候返回-1
                start = char_dict[char]
            char_dict[char] = i  # 向map容器中添加key：字符char；value：i下标
            max_len = max(max_len, i - start)
        return max_len


'''
用2个变量（start, end）分别保存子串的起点和终点。
end自增，直到遇到重复字符为止；
从重复字符出现的位置之后1位重新开始扫描。
'''

File: Algorithms/C++/3-Longest_Substring_Without_Repeating_Characters.cpp
// Runtime: 8 ms, faster than 95.93% of C++ online submissions for Longest Substring Without Repeating Characters.
// Memory Usage: 10.4 MB, less than 67.18% of C++ online submissions for Longest Substring Without Repeating Characters.
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        vector<int> dic(256, -1);
        int max_len = 0, start = -1;
        for (int i = 0; i != s.length(); ++i) {
            if (dic[s[i]] > start)
                start = dic[s[i]];
            dic[s[i]] = i;
            max_len = max(max_len, i - start);
        }
        return max_len;
    }
};


// Runtime: 8 ms, faster than 89.20% of C++ online submissions for Longest Substring Without Repeating Characters.
// Memory Usage: 7.4 MB, less than 82.40% of C++ online submissions for Longest Substring Without Repeating Characters.
/**
 * 最优解
 */
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int ans = 0;
        vector<int> mp(128, 0);
        for (int i = 0, j = 0; j < s.size(); ++j) {
            i = std::max(mp[s[j]], i);
            ans = std::max(ans, j - i + 1);
            mp[s[j]] = j + 1; // 存储的是[1, ..., s.size()]，对应[0, ..., s.size() - 1]
        }
        return ans;
    }
};



// Runtime: 28 ms, faster than 56.00% of C++ online submissions for Longest Substring Without Repeating Characters.
// Memory Usage: 10.8 MB, less than 27.99% of C++ online submissions for Longest Substring Without Repeating Characters.
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int ans = 0,
            left = 0, // left pointer
            right = 0; // right pointer
        std::unordered_set<char> st; // sliding window
        while (left < s.size() && right < s.size()) {
            if (st.find(s[right]) == st.end()) { // contain
                st.insert(s[right]);
                right++;
                ans = std::max(ans, right - left);
            } else { // not contain, 删除left到重复的所有元素
                st.erase(s[left]);
                left++;
            }
            // cout << left << " " << right << endl;
        }
        return ans;
    }
};
